File: quarry/types/proxy.go
// Package types defines core domain types for the Quarry runtime.
package types

import "fmt"

// ProxyProtocol is the allowed proxy protocol.
// Note: socks5 is best-effort with Puppeteer.
type ProxyProtocol string

const (
	ProxyProtocolHTTP   ProxyProtocol = "http"
	ProxyProtocolHTTPS  ProxyProtocol = "https"
	ProxyProtocolSOCKS5 ProxyProtocol = "socks5"
)

// ProxyStrategy is the proxy selection strategy for pools.
type ProxyStrategy string

const (
	ProxyStrategyRoundRobin ProxyStrategy = "round_robin"
	ProxyStrategyRandom     ProxyStrategy = "random"
	ProxyStrategySticky     ProxyStrategy = "sticky"
)

// ProxyStickyScope determines what key is used for sticky assignment.
type ProxyStickyScope string

const (
	ProxyStickyJob    ProxyStickyScope = "job"
	ProxyStickyDomain ProxyStickyScope = "domain"
	ProxyStickyOrigin ProxyStickyScope = "origin"
)

// ProxyEndpoint is a resolved proxy endpoint the executor can dial.
// Emitted by runtime in run requests.
type ProxyEndpoint struct {
	// Protocol is the proxy protocol.
	Protocol ProxyProtocol `json:"protocol" msgpack:"protocol"`
	// Host is the proxy host.
	Host string `json:"host" msgpack:"host"`
	// Port is the proxy port (1-65535).
	Port int `json:"port" msgpack:"port"`
	// Username is the optional username for authentication.
	Username *string `json:"username,omitempty" msgpack:"username,omitempty"`
	// Password is the optional password for authentication.
	Password *string `json:"password,omitempty" msgpack:"password,omitempty"`
}

// Validate validates a proxy endpoint per CONTRACT_PROXY.md hard validation rules.
func (p *ProxyEndpoint) Validate() error {
	// Protocol validation
	switch p.Protocol {
	case ProxyProtocolHTTP, ProxyProtocolHTTPS, ProxyProtocolSOCKS5:
		// valid
	default:
		return fmt.Errorf("invalid protocol %q: must be http, https, or socks5", p.Protocol)
	}

	// Port validation
	if p.Port < 1 || p.Port > 65535 {
		return fmt.Errorf("invalid port %d: must be between 1 and 65535", p.Port)
	}

	// Auth pair validation
	hasUsername := p.Username != nil && *p.Username != ""
	hasPassword := p.Password != nil && *p.Password != ""
	if hasUsername != hasPassword {
		return fmt.Errorf("username and password must be provided together")
	}

	return nil
}

// Redact returns a copy of the endpoint without the password.
func (p *ProxyEndpoint) Redact() ProxyEndpointRedacted {
	return ProxyEndpointRedacted{
		Protocol: p.Protocol,
		Host:     p.Host,
		Port:     p.Port,
		Username: p.Username,
	}
}

// ProxyEndpointRedacted is a proxy endpoint without password.
// Used in run results per CONTRACT_PROXY.md.
type ProxyEndpointRedacted struct {
	Protocol ProxyProtocol `json:"protocol" msgpack:"protocol"`
	Host     string        `json:"host" msgpack:"host"`
	Port     int           `json:"port" msgpack:"port"`
	Username *string       `json:"username,omitempty" msgpack:"username,omitempty"`
}

// ProxySticky is sticky configuration for a proxy pool.
type ProxySticky struct {
	// Scope is the scope for sticky key derivation.
	Scope ProxyStickyScope `json:"scope" msgpack:"scope"`
	// TTLMs is the optional TTL in milliseconds for sticky entries.
	TTLMs *int64 `json:"ttl_ms,omitempty" msgpack:"ttl_ms,omitempty"`
}

// ProxyPool defines a pool and rotation policy.
// Parsed and validated by runtime.
type ProxyPool struct {
	// Name is the pool name (unique identifier).
	Name string `json:"name" msgpack:"name"`
	// Strategy is the selection strategy.
	Strategy ProxyStrategy `json:"strategy" msgpack:"strategy"`
	// Endpoints is the list of available endpoints (must have at least one).
	Endpoints []ProxyEndpoint `json:"endpoints" msgpack:"endpoints"`
	// Sticky is the optional sticky configuration.
	Sticky *ProxySticky `json:"sticky,omitempty" msgpack:"sticky,omitempty"`
}

// Validate validates a proxy pool per CONTRACT_PROXY.md hard validation rules.
func (p *ProxyPool) Validate() error {
	if p.Name == "" {
		return fmt.Errorf("pool name is required")
	}

	switch p.Strategy {
	case ProxyStrategyRoundRobin, ProxyStrategyRandom, ProxyStrategySticky:
		// valid
	default:
		return fmt.Errorf("invalid strategy %q: must be round_robin, random, or sticky", p.Strategy)
	}

	if len(p.Endpoints) == 0 {
		return fmt.Errorf("pool must have at least one endpoint")
	}

	for i, ep := range p.Endpoints {
		if err := ep.Validate(); err != nil {
			return fmt.Errorf("endpoints[%d]: %w", i, err)
		}
	}

	if p.Sticky != nil {
		switch p.Sticky.Scope {
		case ProxyStickyJob, ProxyStickyDomain, ProxyStickyOrigin:
			// valid
		default:
			return fmt.Errorf("invalid sticky scope %q: must be job, domain, or origin", p.Sticky.Scope)
		}

		if p.Sticky.TTLMs != nil && *p.Sticky.TTLMs <= 0 {
			return fmt.Errorf("sticky TTL must be positive")
		}
	}

	return nil
}

// JobProxyRequest is a job-level proxy selection request.
type JobProxyRequest struct {
	// Pool is the pool name to select from.
	Pool string `json:"pool" msgpack:"pool"`
	// Strategy is the optional strategy override.
	Strategy *ProxyStrategy `json:"strategy,omitempty" msgpack:"strategy,omitempty"`
	// StickyKey is the optional sticky key override.
	StickyKey *string `json:"sticky_key,omitempty" msgpack:"sticky_key,omitempty"`
}

// LargePoolThreshold is the number of endpoints above which round_robin
// is discouraged in favor of random.
const LargePoolThreshold = 50

// Warnings returns soft warnings per CONTRACT_PROXY.md.
// These are non-fatal issues that should be surfaced to users.
func (p *ProxyEndpoint) Warnings() []string {
	var warnings []string

	// socks5 is best-effort with Puppeteer
	if p.Protocol == ProxyProtocolSOCKS5 {
		warnings = append(warnings, "socks5 protocol is best-effort with Puppeteer; consider http or https for reliable proxy support")
	}

	return warnings
}

// Warnings returns soft warnings per CONTRACT_PROXY.md.
// These are non-fatal issues that should be surfaced to users.
func (p *ProxyPool) Warnings() []string {
	var warnings []string

	// Very large endpoint lists with round_robin
	if p.Strategy == ProxyStrategyRoundRobin && len(p.Endpoints) > LargePoolThreshold {
		warnings = append(warnings, fmt.Sprintf("pool %q has %d endpoints with round_robin strategy; consider random for large pools", p.Name, len(p.Endpoints)))
	}

	// Check endpoint warnings
	hasSocks5 := false
	for _, ep := range p.Endpoints {
		if ep.Protocol == ProxyProtocolSOCKS5 {
			hasSocks5 = true
			break
		}
	}
	if hasSocks5 {
		warnings = append(warnings, fmt.Sprintf("pool %q contains socks5 endpoints; socks5 is best-effort with Puppeteer", p.Name))
	}

	return warnings
}

File: sdk/src/proxy.ts
/**
 * Proxy validation helpers per CONTRACT_PROXY.md
 */

import type { ProxyEndpoint, ProxyPool, ProxyProtocol, ProxyStrategy } from './types/proxy'

/**
 * Hard validation error that must cause rejection.
 */
export type ProxyValidationError = {
  readonly field: string
  readonly message: string
}

/**
 * Soft warning that should be surfaced but not rejected.
 */
export type ProxyValidationWarning = {
  readonly field: string
  readonly message: string
}

/**
 * Result of proxy validation.
 */
export type ProxyValidationResult = {
  readonly valid: boolean
  readonly errors: readonly ProxyValidationError[]
  readonly warnings: readonly ProxyValidationWarning[]
}

const VALID_PROTOCOLS: readonly ProxyProtocol[] = ['http', 'https', 'socks5']
const VALID_STRATEGIES: readonly ProxyStrategy[] = ['round_robin', 'random', 'sticky']
const MIN_PORT = 1
const MAX_PORT = 65535
const LARGE_POOL_THRESHOLD = 100

function validationError(field: string, message: string): ProxyValidationError {
  return { field, message }
}

function validationWarning(field: string, message: string): ProxyValidationWarning {
  return { field, message }
}

function isPresent(value: string | undefined): boolean {
  return value !== undefined && value !== ''
}

function buildValidationResult(
  errors: ProxyValidationError[],
  warnings: ProxyValidationWarning[]
): ProxyValidationResult {
  return { valid: errors.length === 0, errors, warnings }
}

/**
 * Validate a proxy endpoint per CONTRACT_PROXY hard validation rules.
 *
 * Hard validation (must reject):
 * - host must be a non-empty string
 * - port is within 1-65535
 * - protocol is one of http|https|socks5
 * - username and password must be provided together if either is set
 */
export function validateProxyEndpoint(endpoint: ProxyEndpoint, prefix = ''): ProxyValidationResult {
  const errors: ProxyValidationError[] = []
  const warnings: ProxyValidationWarning[] = []
  const fieldPrefix = prefix ? `${prefix}.` : ''

  // Host validation (required per CONTRACT_PROXY.md)
  if (!endpoint.host || typeof endpoint.host !== 'string') {
    errors.push(validationError(`${fieldPrefix}host`, 'Host is required and must be a non-empty string'))
  }

  // Protocol validation
  if (!VALID_PROTOCOLS.includes(endpoint.protocol)) {
    errors.push(
      validationError(
        `${fieldPrefix}protocol`,
        `Invalid protocol "${endpoint.protocol}". Must be one of: ${VALID_PROTOCOLS.join(', ')}`
      )
    )
  }

  // Port validation
  if (
    typeof endpoint.port !== 'number' ||
    !Number.isInteger(endpoint.port) ||
    endpoint.port < MIN_PORT ||
    endpoint.port > MAX_PORT
  ) {
    errors.push(
      validationError(
        `${fieldPrefix}port`,
        `Invalid port ${endpoint.port}. Must be integer between ${MIN_PORT} and ${MAX_PORT}`
      )
    )
  }

  // Auth pair validation
  if (isPresent(endpoint.username) !== isPresent(endpoint.password)) {
    errors.push(
      validationError(
        `${fieldPrefix}username/${fieldPrefix}password`,
        'Username and password must be provided together'
      )
    )
  }

  // Soft warning: socks5 with Puppeteer
  if (endpoint.protocol === 'socks5') {
    warnings.push(
      validationWarning(`${fieldPrefix}protocol`, 'socks5 usage with Puppeteer is best-effort')
    )
  }

  return buildValidationResult(errors, warnings)
}

/**
 * Validate a proxy pool per CONTRACT_PROXY hard validation rules.
 *
 * Hard validation (must reject):
 * - endpoints length > 0
 * - all endpoints pass validation
 * - strategy is valid
 *
 * Soft warnings:
 * - very large endpoint lists with round_robin
 */
export function validateProxyPool(pool: ProxyPool): ProxyValidationResult {
  const errors: ProxyValidationError[] = []
  const warnings: ProxyValidationWarning[] = []

  // Name validation
  if (!pool.name || typeof pool.name !== 'string') {
    errors.push(validationError('name', 'Pool name is required'))
  }

  // Strategy validation
  if (!VALID_STRATEGIES.includes(pool.strategy)) {
    errors.push(
      validationError(
        'strategy',
        `Invalid strategy "${pool.strategy}". Must be one of: ${VALID_STRATEGIES.join(', ')}`
      )
    )
  }

  // Endpoints validation
  if (!pool.endpoints || pool.endpoints.length === 0) {
    errors.push(validationError('endpoints', 'Pool must have at least one endpoint'))
  } else {
    // Validate each endpoint
    const endpointResults = pool.endpoints.map((endpoint, i) =>
      validateProxyEndpoint(endpoint, `endpoints[${i}]`)
    )
    errors.push(...endpointResults.flatMap((r) => r.errors))
    warnings.push(...endpointResults.flatMap((r) => r.warnings))

    // Soft warning: large pool with round_robin
    if (pool.strategy === 'round_robin' && pool.endpoints.length > LARGE_POOL_THRESHOLD) {
      warnings.push(
        validationWarning(
          'strategy',
          `Large endpoint list (${pool.endpoints.length}) with round_robin strategy; consider using random`
        )
      )
    }
  }

  // Recency window validation
  if (pool.recencyWindow !== undefined) {
    if (
      typeof pool.recencyWindow !== 'number' ||
      !Number.isInteger(pool.recencyWindow) ||
      pool.recencyWindow <= 0
    ) {
      errors.push(validationError('recencyWindow', 'Recency window must be a positive integer'))
    }
    if (pool.strategy !== 'random') {
      warnings.push(
        validationWarning(
          'recencyWindow',
          `Recency window is set but strategy is "${pool.strategy}"; recency window only applies to random selection`
        )
      )
    }
  }

  // Sticky validation
  if (pool.sticky) {
    if (pool.strategy !== 'sticky') {
      warnings.push(
        validationWarning('sticky', 'Sticky configuration provided but strategy is not "sticky"')
      )
    }

    const validScopes = ['job', 'domain', 'origin']
    if (!validScopes.includes(pool.sticky.scope)) {
      errors.push(
        validationError(
          'sticky.scope',
          `Invalid sticky scope "${pool.sticky.scope}". Must be one of: ${validScopes.join(', ')}`
        )
      )
    }

    if (
      pool.sticky.ttlMs !== undefined &&
      (typeof pool.sticky.ttlMs !== 'number' || pool.sticky.ttlMs <= 0)
    ) {
      errors.push(validationError('sticky.ttlMs', 'Sticky TTL must be a positive number'))
    }
  }

  return buildValidationResult(errors, warnings)
}

/**
 * Redact password from a proxy endpoint.
 * Returns a new object without the password field.
 */
export function redactProxyEndpoint<T extends ProxyEndpoint>(endpoint: T): Omit<T, 'password'> {
  const { password: _, ...redacted } = endpoint
  return redacted
}
